src/Interpreters/Cluster.cpp

```diff
@@ -435,18 +435,32 @@ Cluster::Cluster(const Poco::Util::AbstractConfiguration & config,
     }
 
     bool use_static_shard_number = keys_with_static_shard_number == config_keys.size();
+    std::unordered_set<UInt32> used_shard_nums;
     UInt32 current_shard_num = 1;
+
     for (const auto & key : config_keys)
     {
-        if (startsWith(key, "node"))
+        bool shard_with_replicas = startsWith(key, "shard");
+        bool shard_without_replicas = startsWith(key, "node");
+
+        if (!shard_with_replicas && !shard_without_replicas)
+            throw Exception(ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG, "Unknown element in config: {}", key);
+
+        const auto & prefix = config_prefix + key + ((shard_with_replicas) ? ".":  "");
+        const auto weight = config.getInt(prefix + ".weight", default_weight);
+        const auto shard_number = use_static_shard_number ? config.getUInt(prefix + ".shard_number") : current_shard_num;
+        if (use_static_shard_number)
         {
-            /// Shard without replicas.
+            if (!shard_number || used_shard_nums.find(shard_number) != used_shard_nums.end())
```

Comment:
```suggestion
            if (!shard_number || !used_shard_nums.insert(shard_number).second)
```

Then the line 457 can be deleted